### PR TITLE
Implement basic frontend structure

### PR DESCRIPTION
**Pull request by Adda0**
This PR implements the basic frontend structure of Angular project with Bootstrap. The features included are:
- initialization of the project,
- routing and authentication guards,
- division of project into modules (for states, events, board games and users),
- common shared services and models,
- basic authentication support,
- extensible navigation bar and
- basic components for users (login, user profile page, navigation bar account pop-up), events (forms for planning or modifying events, mock components for showing events in a table with options to delete or modify them and event detail page).

The PR does not include final styling of the application and even though some clean-up was already started, there is much more polishing to be done left.

Code at this current state contains a lot of TODOs and FIXMEs, which are meant to be implemented and properly adjusted later when the basic structure is fixed and the proper handling of these is possible because of its dependencies on other PRs.

### CODE REVIEW

File: ClientApp/KachnaOnline/src/app/app.module.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+import { AppRoutingModule } from './app-routing.module';
+import { AppComponent } from './app.component';
+import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
+import { AccountPopupComponent } from './navigation-bar/account-popup/account-popup.component';
+import { NavigationBarComponent } from './navigation-bar/navigation-bar.component';
+import { CurrentEventsComponent } from './events/current-events/current-events.component';
+import { PageNotFoundComponent } from './page-not-found/page-not-found.component';
+import { EventDetailComponent } from './events/event-detail/event-detail.component';
+import {HttpClientModule, HttpRequest} from '@angular/common/http';
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/app.module.ts:17]: Why no spaces

File: ClientApp/KachnaOnline/src/app/app.module.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import { HomeComponent } from './home/home.component';
+import { EventsFromAllComponent } from './events/events-from-all/events-from-all.component';
+import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
+import { ToastrModule } from 'ngx-toastr';
+import { JwtModule } from '@auth0/angular-jwt';
+import { LoggedInContentComponent } from './navigation-bar/account-popup/logged-in-content/logged-in-content.component';
+import { LoggedOutContentComponent } from './navigation-bar/account-popup/logged-out-content/logged-out-content.component';
+import { PlanEventsComponent } from './events/plan-events/plan-events.component';
+import { ForbiddenComponent } from './forbidden/forbidden.component';
+import {UsersModule} from "./users/users.module";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/app.module.ts:29]: Why no spaces

File: ClientApp/KachnaOnline/src/app/app.module.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import { EventsModule} from "./events/events.module";
+import {StatesModule} from "./states/states.module";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/app.module.ts:31]: Why no spaces

File: ClientApp/KachnaOnline/src/app/app.module.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import {BoardGamesModule} from "./board-games/board-games.module";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/app.module.ts:32]: Why no spaces

File: ClientApp/KachnaOnline/src/app/app.module.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    NgbModule,
+    HttpClientModule,
+    FormsModule,
+    ReactiveFormsModule,
+    BrowserAnimationsModule,
+    ToastrModule.forRoot(), // TODO: Change options?
+    JwtModule.forRoot({
+      config: {
+        tokenGetter: tokenGetter,
+        allowedDomains: [environment.baseApiUrlDomain, environment.kisApiUrlDomain/*, 'localhost:5001'*/],
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/app.module.ts:75]: Delete that commented-out localhost maybe?

File: ClientApp/KachnaOnline/src/app/events/events-routing.module.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+// Author: David Chocholatý
+
+import { environment } from '../../environments/environment';
+import { NgModule } from '@angular/core';
+import { RouterModule, Routes } from '@angular/router';
+import {CurrentEventsComponent} from "./current-events/current-events.component";
+import {EventsFromAllComponent} from "./events-from-all/events-from-all.component";
+import {PlanEventsComponent} from "./plan-events/plan-events.component";
+import {EventsManagerGuard} from "./events-manager.guard";
+import {EventDetailComponent} from "./event-detail/event-detail.component";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/events/events-routing.module.ts:11]: Why no spaces

File: ClientApp/KachnaOnline/src/app/models/user.model.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+// user.model.ts
+// Author: David Chocholatý
+
+export class User {
+  name: string = "";
+  email: string = "";
+  nickname: string = "";
+  gamificationConsent: boolean = false;
+  cardNumber: string = "";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/models/user.model.ts:9]: This should rather be called `cardCode` (the validation regex for this is `[A-Z]{5}`).

File: ClientApp/KachnaOnline/src/app/users/admin.guard.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
+// admin.guard.ts
+// Author: David Chocholatý
+
+import {Injectable} from '@angular/core';
+import {ActivatedRouteSnapshot, CanActivate, Router, RouterStateSnapshot, UrlTree} from '@angular/router';
+import {Observable} from 'rxjs';
+import {AuthenticationService} from "../shared/services/authentication.service";
+import {RoleTypes} from "../models/role-types.model";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/users/admin.guard.ts:8]: Why no spaces

File: ClientApp/KachnaOnline/src/app/users/login/login.component.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
+// login.component.ts
+// Author: David Chocholatý
+
+import { AuthenticationService } from '../../shared/services/authentication.service';
+import { JwtHelperService } from '@auth0/angular-jwt';
+import { ToastrService } from 'ngx-toastr';
+import { HttpClient, HttpParams } from '@angular/common/http';
+import { Component, ComponentFactoryResolver, OnInit } from '@angular/core';
+import {ActivatedRoute, Router} from '@angular/router';
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/users/login/login.component.ts:9]: Why no spaces

File: ClientApp/KachnaOnline/src/app/users/user-profile/user-logged-in.guard.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
+// user-logged-in.guard.ts
+// Author: David Chocholatý
+
+import { Injectable } from '@angular/core';
+import {ActivatedRouteSnapshot, CanActivate, Router, RouterStateSnapshot, UrlTree} from '@angular/router';
+import { Observable } from 'rxjs';
+import {AuthenticationService} from "../../shared/services/authentication.service";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/users/user-profile/user-logged-in.guard.ts:7]: Why no spaces

File: ClientApp/KachnaOnline/src/app/users/user-profile/user-profile.component.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
+// user-profile.component.ts
+// Author: David Chocholatý
+
+import { Component, OnInit } from '@angular/core';
+import {AuthenticationService} from "../../shared/services/authentication.service";
+import {NgForm} from "@angular/forms";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/users/user-profile/user-profile.component.ts:6]: Why no spaces

File: ClientApp/KachnaOnline/src/app/users/user-profile/user-profile.component.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+})
+export class UserProfileComponent implements OnInit {
+  public isCardInfoCollapsed: boolean = false;
+
+  constructor(
+    public authenticationService: AuthenticationService,
+  ) { }
+
+  gamificationConsentTooltipText:string = "Tohle je souhlas s gamifikací."
+  cardNumberTooltipText:string = "Tohle je krásná karta."
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/users/user-profile/user-profile.component.ts:21]: Again, `cardCode` would be better than `cardNumber`

File: ClientApp/KachnaOnline/src/app/users/users-routing.module.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+// Author: David Chocholatý
+
+import { NgModule } from '@angular/core';
+import { RouterModule, Routes } from '@angular/router';
+
+import { LoginComponent } from './login/login.component';
+import {UserProfileComponent } from "./user-profile/user-profile.component";
+import {UserLoggedInGuard } from "./user-profile/user-logged-in.guard";
+import {UsersListComponent} from "./users-list/users-list.component";
+import {AdminGuard} from "./admin.guard";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/users/users-routing.module.ts:11]: spaces :(

File: ClientApp/KachnaOnline/src/app/users/users.module.ts [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+import { environment } from '../../environments/environment';
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
+import { UsersRoutingModule } from './users-routing.module';
+import { UsersListComponent } from './users-list/users-list.component';
+import {UserProfileComponent} from "./user-profile/user-profile.component";
+import {FormsModule, NgForm} from "@angular/forms";
+import {NgbModule} from "@ng-bootstrap/ng-bootstrap";
+import {LoginComponent} from "./login/login.component";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/app/users/users.module.ts:12]: spaces :(

File: ClientApp/KachnaOnline/src/environments/environment.prod.ts [hunk 15]
```diff
@@ -0,0 +1,16 @@
+// environment.prod.ts
+// Author: David Chocholatý
+
+import {IEnvironment, IEnvironmentParams} from "./ienvironment";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/environments/environment.prod.ts:4]: spaces

File: ClientApp/KachnaOnline/src/environments/environment.ts [hunk 16]
```diff
@@ -0,0 +1,29 @@
+// environment.ts
+// Author: David Chocholatý
+
+// This file can be replaced during build by using the `fileReplacements` array.
+// `ng build` replaces `environment.ts` with `environment.prod.ts`.
+// The list of file replacements can be found in `angular.json`.
+
+import {IEnvironment, IEnvironmentParams} from "./ienvironment";
```
Comment (ondryaso) [inline ClientApp/KachnaOnline/src/environments/environment.ts:8]: s p a c e s

**Review by ondryaso — CHANGES_REQUESTED** [review]

File: ClientApp/KachnaOnline/angular.json [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+            "polyfills": "src/polyfills.ts",
+            "tsConfig": "tsconfig.app.json",
+            "assets": [
+              "src/favicon.ico",
+              "src/assets"
+            ],
+            "styles": [
+              "node_modules/bootstrap/dist/css/bootstrap.min.css",
+              "src/styles.css",
+              "node_modules/ngx-toastr/toastr.css",
```
Comment (FrNecas) [inline ClientApp/KachnaOnline/angular.json:32]: Invalid JSON syntax

File: ClientApp/KachnaOnline/README.md [hunk 18]
```diff
@@ -0,0 +1,27 @@
+# KachnaOnline
```
Comment (FrNecas) [inline ClientApp/KachnaOnline/README.md:1]: We should later write an actual READMe but lets keep this for now (also some of the commands may be useful for people like me :D )

File: ClientApp/KachnaOnline/src/app/app.module.ts (hunk 1 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/app.module.ts]: Fixed.

File: ClientApp/KachnaOnline/src/app/app.module.ts (hunk 2 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/app.module.ts]: Fixed.

File: ClientApp/KachnaOnline/src/app/app.module.ts (hunk 3 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/app.module.ts]: Fixed.

File: ClientApp/KachnaOnline/src/app/app.module.ts (hunk 4 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/app.module.ts]: Fixed.

File: ClientApp/KachnaOnline/src/app/app.module.ts (hunk 5 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/app.module.ts:74]: Sure. I have missed this when cleaning-up.

File: ClientApp/KachnaOnline/src/app/events/events-routing.module.ts (hunk 6 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/events/events-routing.module.ts]: Fixed.

File: ClientApp/KachnaOnline/src/app/users/admin.guard.ts (hunk 8 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/users/admin.guard.ts]: Fixed.

File: ClientApp/KachnaOnline/src/app/users/login/login.component.ts (hunk 9 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/users/login/login.component.ts]: Fixed.

File: ClientApp/KachnaOnline/src/app/users/user-profile/user-logged-in.guard.ts (hunk 10 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/users/user-profile/user-logged-in.guard.ts]: Fixed.

File: ClientApp/KachnaOnline/src/app/users/user-profile/user-profile.component.ts (hunk 11 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/users/user-profile/user-profile.component.ts]: Fixed.

File: ClientApp/KachnaOnline/src/app/users/users-routing.module.ts (hunk 13 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/users/users-routing.module.ts]: Fixed.

File: ClientApp/KachnaOnline/src/app/users/users.module.ts (hunk 14 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/users/users.module.ts]: Fixed.

File: ClientApp/KachnaOnline/src/environments/environment.prod.ts (hunk 15 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/environments/environment.prod.ts]: Fixed.

File: ClientApp/KachnaOnline/src/environments/environment.ts (hunk 16 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/environments/environment.ts]: Fixed.

File: ClientApp/KachnaOnline/src/app/models/user.model.ts (hunk 7 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/models/user.model.ts:9]: Thank you. I was intending to ask you about that. I will add the constraints now, then.

**Review by FrNecas — APPROVED** [review]
LGTM as far as I can tell other than these few things. Lets merge ASAP and improve further down the line, this looks like a good baseline.

File: ClientApp/KachnaOnline/README.md (hunk 18 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/README.md:1]: It has a README? That was auto-generated. I did not even know it created one. Angular CLI is trying to automatically initialize a Git repo when creating a new project, but when it realizes the path is already in a Git repo, it stops with it, but, as we can see here, leaves behind what has been already created for the repo... Should I remove the file now entirely, or keep it and fill it later?

File: ClientApp/KachnaOnline/angular.json (hunk 17 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/angular.json:32]: Ah, yes. I will fix that.

File: ClientApp/KachnaOnline/src/app/users/user-profile/user-profile.component.ts (hunk 12 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/src/app/users/user-profile/user-profile.component.ts:21]: Good point. Fixing.

File: ClientApp/KachnaOnline/README.md (hunk 18 above)
Comment (Adda0) [reply ClientApp/KachnaOnline/README.md:1]: Decided to keep for now and update its content later.

**Pull request merged**